overlord/snapstate/snapstate.go

```diff
@@ -1201,14 +1203,47 @@ func Install(ctx context.Context, st *state.State, name string, opts *RevisionOp
 	return InstallWithDeviceContext(ctx, st, name, opts, userID, flags, nil, "")
 }
 
+type snapInfoForInstall func(DeviceContext, *RevisionOptions) (si *snap.Info, snapPath, redirectChannel string, e error)
+
 // InstallWithDeviceContext returns a set of tasks for installing a snap.
-// It will query for the snap with the given deviceCtx.
+// It will query the store for the snap with the given deviceCtx.
 // Note that the state must be locked by the caller.
 //
 // The returned TaskSet will contain a LastBeforeLocalModificationsEdge
 // identifying the last task before the first task that introduces system
 // modifications.
 func InstallWithDeviceContext(ctx context.Context, st *state.State, name string, opts *RevisionOptions, userID int, flags Flags, deviceCtx DeviceContext, fromChange string) (*state.TaskSet, error) {
+	logger.Debugf("installing with device context %s", name)
+	snapInstallInfo := func(dc DeviceContext, ro *RevisionOptions) (si *snap.Info, snapPath, redirectChannel string, e error) {
+		sar, err := installInfo(ctx, st, name, ro, userID, flags, dc)
+		if err != nil {
+			return nil, "", "", err
+		}
+		return sar.Info, "", sar.RedirectChannel, nil
+	}
+	return installWithDeviceContext(st, name, opts, userID, flags, deviceCtx, fromChange, snapInstallInfo)
+}
+
+// InstallPathWithDeviceContext returns a set of tasks for installing a local snap.
+// It will query the store for the snap with the given deviceCtx.
+// Note that the state must be locked by the caller.
+//
+// The returned TaskSet will contain a LastBeforeLocalModificationsEdge
+// identifying the last task before the first task that introduces system
+// modifications.
+func InstallPathWithDeviceContext(st *state.State, name string, opts *RevisionOptions, userID int, flags Flags, deviceCtx DeviceContext, fromChange string, si *snap.SideInfo, path string) (*state.TaskSet, error) {
```

Comment:
this doesn't seem tested

Reply:
the placement of si and path in the arguments is odd, it should be closer to InstallPath

Reply:
Changed